Netkan/Services/CachingHttpService.cs

```diff
@@ -0,0 +1,141 @@
+
+            _curl = new CurlEasy { UserAgent = CKAN.Net.UserAgentString };
+
+            var caBundle = ResolveCurlCaBundle();
+            if (caBundle != null)
+            {
+                _curl.CaInfo = caBundle;
+            }
+        }
+
+        public string DownloadPackage(Uri url, string identifier)
+        {
+            EnsureNotDisposed();
+
+            return _cache.GetCachedFilename(url) ?? 
+                _cache.Store(url, CKAN.Net.Download(url), string.Format("netkan-{0}", identifier), move: true);
+        }
+
+        public string DownloadText(Uri url)
+        {
+            EnsureNotDisposed();
+
+            Log.DebugFormat("About to download {0}", url);
+
+            var content = string.Empty;
+
+            _curl.Url = url.ToString();
+            _curl.WriteData = null;
+            _curl.WriteFunction = delegate(byte[] buf, int size, int nmemb, object extraData)
+            {
+                content += Encoding.UTF8.GetString(buf);
+                return size * nmemb;
+            };
+
+            var result = _curl.Perform();
+
+            if (result != CurlCode.Ok)
+            {
+                throw new Exception("Curl download failed with error " + result);
+            }
+
+            Log.DebugFormat("Download from {0}:\n\n{1}", url, content);
+
+            return content;
+        }
+
+        /// <summary>
+        /// Resolves the location of the cURL CA bundle file to use.
+        /// </summary>
+        /// <returns>The absolute file path to the bundle file or null if none is found.</returns>
+        private static string ResolveCurlCaBundle()
+        {
+            const string caBundleFileName = "curl-ca-bundle.crt";
+            const string ckanSubDirectoryName = "CKAN";
+
+            var bundle = new[]
+            {
+                // Working Directory
+                Environment.CurrentDirectory,
+
+                // Executable Directory
+                Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location),
+
+                // %LOCALAPPDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), ckanSubDirectoryName),
+
+                // %APPDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), ckanSubDirectoryName),
+
+                // %PROGRAMDATA%/CKAN
+                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), ckanSubDirectoryName),
+            }
+            .Select(i => Path.Combine(i, caBundleFileName))
+            .FirstOrDefault(File.Exists);
+
+            Log.InfoFormat("Using curl-ca bundle: {0}",bundle ?? "(none)");
+
+            return bundle;
+        }
+    }
+
+    internal sealed partial class CachingHttpService
+    {
+        private bool _isDisposed;
+
+        ~CachingHttpService()
+        {
+            Dispose(disposing: false);
```

Comment:
> The destructor will be called if the object is about to be garbage collected without someone previously calling `Dispose()` (usually via a `using` statement)

Time for me to learn about the GC some more. Am I correct the finaliser runs *after* dispose, but as a final pass before the object is completely destroyed?

If that's the case, and we're *not* cleaning up our managed resources, the assumption is that by the time we reach the finaliser then `.Dispose()` will have already been called, yes? In that case I fear I may be misunderstanding the pattern; I would expect us to check to see if our managed resources are still around and dispose of them if so, making sure we don't accidentally leak if our object falls out of scope and wasn't created with `using`.